File: src/main/java/es/developer/achambi/coreframework/threading/LifecycleResponseHandler.java
package es.developer.achambi.coreframework.threading;

import android.arch.lifecycle.Lifecycle;

public class LifecycleResponseHandler<T> extends ResponseHandlerDecorator<T> {
    private Lifecycle lifecycle;

    public LifecycleResponseHandler( Lifecycle lifecycle,  ResponseHandler responseHandler) {
        super(responseHandler);
        this.lifecycle = lifecycle;
    }

    @Override
    public void onSuccess(Response<T> response) {
        if( lifecycle.getCurrentState().isAtLeast( Lifecycle.State.CREATED ) ) {
            super.onSuccess(response);
        }
    }

    @Override
    public void onError(Error error) {
        if( lifecycle.getCurrentState().isAtLeast( Lifecycle.State.CREATED ) ) {
            super.onError(error);
        }
    }
}

File: src/main/java/es/developer/achambi/coreframework/threading/LifecycleResponseHandler.kt
package es.developer.achambi.coreframework.threading

import androidx.lifecycle.Lifecycle

class LifecycleResponseHandler<T>
(private val lifecycle: Lifecycle, responseHandler: ResponseHandler<T>)
    : ResponseHandlerDecorator<T>(responseHandler) {

    override fun onSuccess(response: T) {
        if (lifecycle.currentState.isAtLeast(Lifecycle.State.STARTED)) {
            super.onSuccess(response)
        }
    }

    override fun onError(error: CoreError) {
        if (lifecycle.currentState.isAtLeast(Lifecycle.State.STARTED)) {
            super.onError(error)
        }
    }
}
